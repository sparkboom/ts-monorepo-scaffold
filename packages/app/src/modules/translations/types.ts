export type Texts = object;
export type Locale = 'de' | 'es' | 'es-XL' | 'en' | 'fr' | 'fr-CA' | 'it' | 'ja' | 'pt' | 'qaT';
export type TextsDict = {
  [key: string]: string;
};
export type TextsCollection = {
  [locale in Locale]: TextsDict | undefined;
};
export interface TextEntry {
  key: string;
  values?: {
    [k: string]: string | number;
  };
}
