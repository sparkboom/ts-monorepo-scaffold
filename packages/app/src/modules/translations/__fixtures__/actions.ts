export const emptyAction = { type: '', payload: {}, correlationId: '' };
