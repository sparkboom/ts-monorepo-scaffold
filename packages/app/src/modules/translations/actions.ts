/* eslint-disable import/prefer-default-export */
import { Texts, Locale, TextsDict } from './types';

// Types
type SetTranslationsPayload = {
  [key in Locale]?: TextsDict;
};
type SetTranslations = (locale: Locale, dict: Texts) => AppAction<'TRANSLATIONS__SET', SetTranslationsPayload>;

// Actions
type LoadTranslations = (cultureCode: Locale) => AppAction<'TRANSLATIONS__LOAD', { cultureCode: Locale }>;
export const loadTranslations: LoadTranslations = cultureCode => ({
  type: 'TRANSLATIONS__LOAD',
  payload: {
    cultureCode,
  },
});

export const setTranslations: SetTranslations = (locale, dict) => ({
  type: 'TRANSLATIONS__SET',
  payload: {
    [locale]: dict,
  },
});

export const useTranslations = locale => ({
  type: 'TRANSLATIONS__USE',
  payload: {
    locale,
  },
});

// TODO: Use idiomatic fail, warn actions from logging
type LoadTranslationsFail = (message: string) => AppAction<'TRANSLATIONS__LOAD_FAIL', { message: string }>;
export const loadTranslationsFail: LoadTranslationsFail = message => ({
  type: 'TRANSLATIONS__LOAD_FAIL',
  payload: {
    message,
  },
});

type LoadTranslationsWarn = (message: string) => AppAction<'TRANSLATIONS__LOAD_WARN', { message: string }>;
export const loadTranslationsWarn: LoadTranslationsWarn = message => ({
  type: 'TRANSLATIONS__LOAD_WARN',
  payload: {
    message,
  },
});

type UseTranslationsFail = (message: string) => AppAction<'TRANSLATIONS__USE_FAIL', { message: string }>;
export const useTranslationsFail: UseTranslationsFail = message => ({
  type: 'TRANSLATIONS__USE_FAIL',
  payload: {
    message,
  },
});
