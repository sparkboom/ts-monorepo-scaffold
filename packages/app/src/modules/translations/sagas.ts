import { takeEvery, takeLatest, call, select, apply, PutEffect, SelectEffect, CallEffect, ForkEffect } from 'redux-saga/effects';
import flat from 'flat';
import { isEmpty } from 'lodash-es';
import { updateIntl } from 'react-intl-redux';
import IntlMessageFormat from 'intl-messageformat';
import { Action } from 'redux';
import { ClientWebApi } from '@adstudio-leadcap/api';
import { setTranslations } from './actions';
import { fail, warn } from '../Logging/actions';
import { putContext } from '../Logging/effects';
import { getWebApi } from '../Api/sagas';
import {
  substituteTranslationsSelector,
  fallbackLocaleSelector,
  availableLocalesSelector,
  currentLocaleSelector,
  downloadedLocalesSelector,
  textsSelector,
  messagesSelector,
} from './selectors';
import { Locale, TextsDict, TextsCollection, TextEntry } from './types';

// NOTE:  Official cultures translated in Marketing Cloud are de, en, es-xl, es, fr-ca, fr, it, ja, pt,  and dev language

// Constants
const LOG_CTX = 'CONTENT.SAGAS';

// Types
type MatchLocaleSaga = Generator<SelectEffect | PutEffect<Action<any>>, any, unknown>;
type LoadTranslationsSaga = Generator<SelectEffect | CallEffect<any> | PutEffect<Action<any>>, void, unknown>;
type UseTranslationsSaga = Generator<SelectEffect | PutEffect<Action<any>>, void, unknown>;
type GetTextSaga = Generator<SelectEffect | PutEffect<Action<any>>, string[], any>;
type TranslateSaga = Generator<ForkEffect<never>, void, unknown>;

// Sagas
export function* matchLocaleSaga(action: Action<any>): MatchLocaleSaga {
  const put = putContext(action, LOG_CTX);

  // Prep culture code
  const cultureCode = `${action.payload.cultureCode}`.toLowerCase() as Locale;
  if (!cultureCode || cultureCode.match(/^[a-z]{2}(-[a-z]{2})?$/g) === null) {
    yield put.fail(`Could not fetch translations as culture code '${cultureCode}' is invalid`, { cultureCode });
    return null;
  }

  // Get available locales
  const availableLocales = (yield select(availableLocalesSelector)) as Locale[];
  if (isEmpty(availableLocales)) {
    yield put.fail(`There are no available locales ${availableLocales}`, { availableLocales });
    return null;
  }

  // Return if a direct match is found
  if (availableLocales.includes(cultureCode)) {
    return cultureCode;
  }

  // Return if language family matches
  const cultureCodes = cultureCode.split('-');
  if (cultureCodes.length > 1 && availableLocales.includes(cultureCodes[0])) {
    yield put.warn(`The desired locale '${cultureCode}' is not available, but a proposed match was found '${cultureCodes[0]}'.`, {
      availableLocales,
      cultureCode,
    });
    return cultureCodes[0];
  }

  // Otherwise propose the fallback language
  const fallbackLocale = yield select(fallbackLocaleSelector);
  yield put.warn(`The desired locale '${cultureCode}' is not available, fallback locale '${fallbackLocale}' proposed instead`, { cultureCode, fallbackLocale });

  return fallbackLocale;
}

export function* loadTranslationsSaga(action: Action<any>): LoadTranslationsSaga {
  const put = putContext(action, LOG_CTX);

  const { cultureCode } = action.payload;

  const matchedLocale = (yield call(matchLocaleSaga, action)) as Locale;
  if (matchedLocale === null) {
    yield put.fail(`No locale matched the culture code '${cultureCode}'`);
    return;
  }

  const currentLocale = yield select(currentLocaleSelector);
  if (matchedLocale === currentLocale) {
    yield put.warn(`The desired locale '${matchedLocale}' is currently being used`);
    return;
  }

  const downloadedLocales = (yield select(downloadedLocalesSelector)) as Locale[];
  if (downloadedLocales && downloadedLocales.includes(matchedLocale)) {
    yield put.warn(`The desired locale '${matchedLocale}' has already been downloaded`);
    return;
  }

  const webApi = (yield call(getWebApi, action)) as ClientWebApi;
  if (!webApi) {
    return;
  }
  const translations = (yield apply(webApi, webApi[`${matchedLocale}Texts`])) as Record<string, any> | undefined;
  if (isEmpty(translations)) {
    yield put.fail(`Empty translation set was returned for locale '${matchedLocale}'. Was the API call successful.`);
    return;
  }

  const substituteTexts = (yield select(substituteTranslationsSelector)) as Record<string, any>;
  const flatTexts = flat(translations);
  const texts = {
    ...substituteTexts,
    ...flatTexts,
  };
  yield put(setTranslations(matchedLocale, texts));
}

export function* useTranslationSaga(action: AppAction<string, any>): UseTranslationsSaga {
  const put = putContext(action, LOG_CTX);

  const { locale } = action.payload;
  const texts = (yield select(textsSelector)) as TextsCollection;
  const messages = texts[locale];
  if (!messages) {
    yield put.fail(`The locale '${locale}' has not been loaded`);
    return;
  }
  yield put(updateIntl({ locale, messages }));
}

export function* getTextSaga(texts: TextEntry[], action: Action<any>): GetTextSaga {
  const messageDict = (yield select(messagesSelector)) as TextsDict;
  const locale = (yield select(currentLocaleSelector)) as Locale;
  const unmatchedTexts: string[] = [];
  const messages = texts.map(text => {
    const textTemplate = messageDict[text.key];
    if (!textTemplate) {
      unmatchedTexts.push(text.key);
      return '';
    }
    const formattedMessage = new IntlMessageFormat(textTemplate, locale).format(text.values || {});
    return formattedMessage;
  });
  if (!isEmpty(unmatchedTexts)) {
    yield putContext(action, LOG_CTX).logWarn(`${unmatchedTexts.length} text entry key(s) were not matched the texts dictionary with locale ${locale}`, {
      unmatchedTexts,
    });
  }

  return messages;
}

function* translationsSaga(): TranslateSaga {
  yield takeEvery('TRANSLATIONS__USE', useTranslationSaga);
  yield takeLatest('TRANSLATIONS__LOAD', loadTranslationsSaga);
}

export default translationsSaga;
