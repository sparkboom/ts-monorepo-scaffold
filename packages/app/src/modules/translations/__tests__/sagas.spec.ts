import { expectSaga } from 'redux-saga-test-plan';
import * as matchers from 'redux-saga-test-plan/matchers';
import * as sagas from '../sagas';
import * as selectors from '../selectors';
import * as actions from '../actions';
import { correlate } from '../../../middleware/correlate-redux-middleware';
import { Locale } from '../types';
import { getWebApi } from '../../Api/sagas';

// Mocks
jest.mock('@adstudio-leadcap/api');

// Tests
describe('Translations > Sagas', () => {
  describe('matchLocaleSaga', () => {
    it('should fail with invalid culture code', () => {
      // Arrange
      const FAKE_INVALID_LOCALE_CODE = '-en' as Locale;
      const FAKE_ACTION = correlate(actions.loadTranslations(FAKE_INVALID_LOCALE_CODE), 'FAKE_CORRELATION_ID');
      const EXPECTED_EFFECT = {
        payload: {
          message: "Could not fetch translations as culture code '-en' is invalid",
        },
        correlationId: 'FAKE_CORRELATION_ID',
        type: 'TRANSLATIONS__LOAD_FAIL',
      };
      // Act
      // Assert
      return expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .put.like({ action: EXPECTED_EFFECT })
        .run();
    });

    it('should fail if there are no available locales', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__LOAD_FAIL',
        payload: {
          message: 'There are no available locales ',
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      const FAKE_AVAILABLE_LOCALES: Locale[] = [];
      // Act
      // Assert
      return expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.availableLocalesSelector), FAKE_AVAILABLE_LOCALES]])
        .put.like({ action: EXPECTED_EFFECT })
        .run();
    });

    it('should return the locale if a complete match is found', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_AVAILABLE_LOCALES: Locale[] = ['en', 'es'];
      // Act
      // Assert
      return expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.availableLocalesSelector), FAKE_AVAILABLE_LOCALES]])
        .returns('es')
        .run();
    });

    it("should return the primary matching locale 'es' if no direct one is found 'es-xl'", () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es-xl'), 'FAKE_CORRELATION_ID');
      const FAKE_AVAILABLE_LOCALES: Locale[] = ['es'];
      // Act
      // Assert
      return expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.availableLocalesSelector), FAKE_AVAILABLE_LOCALES]])
        .returns('es')
        .run();
    });

    it('should return the fallback locale if no locale has been matched', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es-xl'), 'FAKE_CORRELATION_ID');
      const FAKE_AVAILABLE_LOCALES: Locale[] = ['en', 'de'];
      const FAKE_FALLBACK_LOCALE: Locale = 'en';
      // Act
      // Assert
      return expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .provide([
          [matchers.select(selectors.availableLocalesSelector), FAKE_AVAILABLE_LOCALES],
          [matchers.select(selectors.fallbackLocaleSelector), FAKE_FALLBACK_LOCALE],
        ])
        .returns('en')
        .run();
    });

    it('should match snapshot when a matching locale is found', async () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_AVAILABLE_LOCALES: Locale[] = ['en', 'es'];

      // Act
      const result = await expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.availableLocalesSelector), FAKE_AVAILABLE_LOCALES]])
        .run();

      // Assert
      expect(result.toJSON()).toMatchSnapshot();
    });

    it('should match snapshot when a fallback locale is used', async () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es-xl'), 'FAKE_CORRELATION_ID');
      const FAKE_AVAILABLE_LOCALES: Locale[] = ['en', 'de'];
      const FAKE_FALLBACK_LOCALE: Locale = 'en';

      // Act
      const result = await expectSaga(sagas.matchLocaleSaga, FAKE_ACTION)
        .provide([
          [matchers.select(selectors.availableLocalesSelector), FAKE_AVAILABLE_LOCALES],
          [matchers.select(selectors.fallbackLocaleSelector), FAKE_FALLBACK_LOCALE],
        ])
        .run();

      // Assert
      expect(result.toJSON()).toMatchSnapshot();
    });
  });

  describe('loadTranslationsSaga', () => {
    it('should fail if no available locale matched the culture code', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__LOAD_FAIL',
        payload: {
          message: "No locale matched the culture code 'es'",
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([[matchers.call.fn(sagas.matchLocaleSaga), null]])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should warn if the desired locale is already being used', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__LOAD_WARN',
        payload: {
          message: "The desired locale 'es' is currently being used",
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([[matchers.call.fn(sagas.matchLocaleSaga), 'es'], [matchers.select(selectors.currentLocaleSelector), 'es']])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should warn if the desired locale has already been downloaded', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__LOAD_WARN',
        payload: {
          message: "The desired locale 'es' has already been downloaded",
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([
          [matchers.call.fn(sagas.matchLocaleSaga), 'es'],
          [matchers.select(selectors.currentLocaleSelector), 'en'],
          [matchers.select(selectors.downloadedLocalesSelector), ['en', 'es']],
        ])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should download the localized texts if they have not already been downloaded', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_WEB_API = {
        esTexts: jest.fn(),
      };
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__LOAD_FAIL',
        payload: {
          message: "Empty translation set was returned for locale 'es'. Was the API call successful.",
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([
          [matchers.call.fn(sagas.matchLocaleSaga), 'es'],
          [matchers.select(selectors.currentLocaleSelector), 'en'],
          [matchers.select(selectors.downloadedLocalesSelector), ['en']],
          [matchers.call.fn(getWebApi), FAKE_WEB_API],
          [matchers.call.fn(FAKE_WEB_API.esTexts), null],
        ])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should fail if not translations are returned', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_WEB_API = {
        esTexts: jest.fn(),
      };
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__LOAD_FAIL',
        payload: {
          message: "Empty translation set was returned for locale 'es'. Was the API call successful.",
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([
          [matchers.call.fn(sagas.matchLocaleSaga), 'es'],
          [matchers.select(selectors.currentLocaleSelector), 'en'],
          [matchers.select(selectors.downloadedLocalesSelector), ['en']],
          [matchers.call.fn(getWebApi), FAKE_WEB_API],
          [matchers.call.fn(FAKE_WEB_API.esTexts), null],
        ])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should store newly downloaded localized texts', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_WEB_API = {
        esTexts: jest.fn(),
      };
      const FAKE_TEXTS = {
        label: 'bot√≥n cancelar',
      };
      const FAKE_SUBSTITUTION_TEXTS = {
        label: 'Cancel Button',
      };
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__SET',
        payload: {
          es: FAKE_TEXTS,
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([
          [matchers.call.fn(sagas.matchLocaleSaga), 'es'],
          [matchers.select(selectors.currentLocaleSelector), 'en'],
          [matchers.select(selectors.downloadedLocalesSelector), ['en']],
          [matchers.call.fn(getWebApi), FAKE_WEB_API],
          [matchers.call.fn(FAKE_WEB_API.esTexts), FAKE_TEXTS],
          [matchers.select(selectors.substituteTranslationsSelector), FAKE_SUBSTITUTION_TEXTS],
        ])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should store newly downloaded localized texts and use subtitutions if necessary', () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_WEB_API = {
        esTexts: jest.fn(),
      };
      const FAKE_TEXTS = {
        label: 'bot√≥n cancelar',
      };
      const FAKE_SUBSTITUTION_TEXTS = {
        label: 'Cancel Button',
        description: 'The is the subtitution for the description value',
      };
      const EXPECTED_EFFECT = {
        type: 'TRANSLATIONS__SET',
        payload: {
          es: {
            label: 'bot√≥n cancelar',
            description: 'The is the subtitution for the description value',
          },
        },
        correlationId: 'FAKE_CORRELATION_ID',
      };
      // Act
      // Assert
      return expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([
          [matchers.call.fn(sagas.matchLocaleSaga), 'es'],
          [matchers.select(selectors.currentLocaleSelector), 'en'],
          [matchers.select(selectors.downloadedLocalesSelector), ['en']],
          [matchers.call.fn(getWebApi), FAKE_WEB_API],
          [matchers.call.fn(FAKE_WEB_API.esTexts), FAKE_TEXTS],
          [matchers.select(selectors.substituteTranslationsSelector), FAKE_SUBSTITUTION_TEXTS],
        ])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should match snapshot', async () => {
      // Arrange
      const FAKE_ACTION = correlate(actions.loadTranslations('es'), 'FAKE_CORRELATION_ID');
      const FAKE_WEB_API = {
        esTexts: jest.fn(),
      };
      const FAKE_TEXTS = {
        label: 'bot√≥n cancelar',
      };
      const FAKE_SUBSTITUTION_TEXTS = {
        label: 'Cancel Button',
      };

      // Act
      const result = await expectSaga(sagas.loadTranslationsSaga, FAKE_ACTION)
        .provide([
          [matchers.call.fn(sagas.matchLocaleSaga), 'es'],
          [matchers.select(selectors.currentLocaleSelector), 'en'],
          [matchers.select(selectors.downloadedLocalesSelector), ['en']],
          [matchers.call.fn(getWebApi), FAKE_WEB_API],
          [matchers.call.fn(FAKE_WEB_API.esTexts), FAKE_TEXTS],
          [matchers.select(selectors.substituteTranslationsSelector), FAKE_SUBSTITUTION_TEXTS],
        ])
        .run();

      // Assert
      expect(result.toJSON()).toMatchSnapshot();
    });
  });

  describe('useTranslationSaga', () => {
    it('should fail if the locale does not match', () => {
      // Arrange
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_ACTION = correlate(actions.useTranslations(FAKE_LOCALE_CODE), 'FAKE_CORRELATION_ID');
      const FAKE_TEXTS_COLLECTION = {
        en: undefined,
      };
      const EXPECTED_EFFECT = {
        payload: {
          message: "The locale 'en' has not been loaded",
        },
        correlationId: 'FAKE_CORRELATION_ID',
        type: 'TRANSLATIONS__USE_FAIL',
      };
      // Act
      // Assert
      return expectSaga(sagas.useTranslationSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.textsSelector), FAKE_TEXTS_COLLECTION]])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should apply locale and messages if there is a match', () => {
      // Arrange
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_ACTION = correlate(actions.useTranslations(FAKE_LOCALE_CODE), 'FAKE_CORRELATION_ID');
      const FAKE_TEXTS_COLLECTION = {
        en: {},
      };
      const EXPECTED_EFFECT = {
        payload: {
          locale: 'en',
          messages: {},
        },
        correlationId: 'FAKE_CORRELATION_ID',
        type: '@@intl/UPDATE',
      };
      // Act
      // Assert
      return expectSaga(sagas.useTranslationSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.textsSelector), FAKE_TEXTS_COLLECTION]])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(undefined)
        .run();
    });

    it('should match snapshot', async () => {
      // Arrange
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_ACTION = correlate(actions.useTranslations(FAKE_LOCALE_CODE), 'FAKE_CORRELATION_ID');
      const FAKE_TEXTS_COLLECTION = {
        en: {},
      };
      // Act
      const result = await expectSaga(sagas.useTranslationSaga, FAKE_ACTION)
        .provide([[matchers.select(selectors.textsSelector), FAKE_TEXTS_COLLECTION]])
        .run();

      // Assert
      expect(result.toJSON()).toMatchSnapshot();
    });
  });

  describe('getTextSaga', () => {
    it('should get single text item from texts dictionary', () => {
      // Arrange
      const FAKE_ACTION = { type: 'FAKE_ACTION' };
      const FAKE_TEXT_ENTRY = [
        {
          key: 'label',
          values: {},
        },
      ];
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_MESSAGES = {
        label: 'label',
      };
      const EXPECTED_STRINGS = ['[en]label {}'];
      // Act
      // Assert
      return expectSaga(sagas.getTextSaga, FAKE_TEXT_ENTRY, FAKE_ACTION)
        .provide([[matchers.select(selectors.messagesSelector), FAKE_MESSAGES], [matchers.select(selectors.currentLocaleSelector), FAKE_LOCALE_CODE]])
        .returns(EXPECTED_STRINGS)
        .run();
    });

    it('should get multiple text items from texts dictionary', () => {
      // Arrange
      const FAKE_ACTION = { type: 'FAKE_ACTION' };
      const FAKE_TEXT_ENTRY = [
        {
          key: 'label',
          values: {},
        },
        {
          key: 'label2',
          values: {},
        },
      ];
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_MESSAGES = {
        label: 'label',
        label2: 'label dos',
      };
      const EXPECTED_STRINGS = ['[en]label {}', '[en]label dos {}'];
      // Act
      // Assert
      return expectSaga(sagas.getTextSaga, FAKE_TEXT_ENTRY, FAKE_ACTION)
        .provide([[matchers.select(selectors.messagesSelector), FAKE_MESSAGES], [matchers.select(selectors.currentLocaleSelector), FAKE_LOCALE_CODE]])
        .returns(EXPECTED_STRINGS)
        .run();
    });

    it('should log warning if no text found', () => {
      // Arrange
      const FAKE_ACTION = { type: 'FAKE_ACTION' };
      const FAKE_TEXT_ENTRY = [
        {
          key: 'label',
          values: {},
        },
        {
          key: 'label2',
          values: {},
        },
      ];
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_MESSAGES = {
        label: 'label',
      };
      const EXPECTED_STRINGS = ['[en]label {}', ''];
      const EXPECTED_EFFECT = {
        type: 'LOGGING__SUBMIT',
        payload: {
          level: 'WARN',
          context: 'CONTENT.SAGAS',
          message: '1 text entry key(s) were not matched the texts dictionary with locale en',
        },
      };
      // Act
      // Assert
      return expectSaga(sagas.getTextSaga, FAKE_TEXT_ENTRY, FAKE_ACTION)
        .provide([[matchers.select(selectors.messagesSelector), FAKE_MESSAGES], [matchers.select(selectors.currentLocaleSelector), FAKE_LOCALE_CODE]])
        .put.like({ action: EXPECTED_EFFECT })
        .returns(EXPECTED_STRINGS)
        .run();
    });

    it('should match snapshot', async () => {
      // Arrange
      const FAKE_ACTION = { type: 'FAKE_ACTION' };
      const FAKE_TEXT_ENTRY = [
        {
          key: 'label',
          values: {},
        },
        {
          key: 'label2',
          values: {},
        },
      ];
      const FAKE_LOCALE_CODE = 'en' as Locale;
      const FAKE_MESSAGES = {
        label: 'label',
        label2: 'label dos',
      };
      // Act
      // Assert
      const result = await expectSaga(sagas.getTextSaga, FAKE_TEXT_ENTRY, FAKE_ACTION)
        .provide([[matchers.select(selectors.messagesSelector), FAKE_MESSAGES], [matchers.select(selectors.currentLocaleSelector), FAKE_LOCALE_CODE]])
        .run();

      // Assert
      expect(result.toJSON()).toMatchSnapshot();
    });
  });
});
