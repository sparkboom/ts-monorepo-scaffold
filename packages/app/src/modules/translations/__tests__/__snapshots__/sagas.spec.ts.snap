// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translations > Sagas getTextSaga should match snapshot 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Translations > Sagas loadTranslationsSaga should match snapshot 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "correlationId": "FAKE_CORRELATION_ID",
            "payload": Object {
              "cultureCode": "es",
            },
            "type": "TRANSLATIONS__LOAD",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/matchLocaleSaga",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          Object {
            "correlationId": "FAKE_CORRELATION_ID",
            "payload": Object {
              "cultureCode": "es",
            },
            "type": "TRANSLATIONS__LOAD",
          },
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/getWebApi",
      },
      "type": "CALL",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "context": Object {
          "esTexts": "@@redux-saga-test-plan/json/function/mockConstructor",
        },
        "fn": "@@redux-saga-test-plan/json/function/mockConstructor",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "FAKE_CORRELATION_ID",
          "payload": Object {
            "es": Object {
              "label": "bot√≥n cancelar",
            },
          },
          "type": "TRANSLATIONS__SET",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Translations > Sagas matchLocaleSaga should match snapshot when a fallback locale is used 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "FAKE_CORRELATION_ID",
          "payload": Object {
            "details": Object {
              "cultureCode": "es-xl",
              "fallbackLocale": "en",
            },
            "message": "The desired locale 'es-xl' is not available, fallback locale 'en' proposed instead",
            "relatedAction": Object {
              "correlationId": "FAKE_CORRELATION_ID",
              "payload": Object {
                "cultureCode": "es-xl",
              },
              "type": "TRANSLATIONS__LOAD",
            },
          },
          "type": "TRANSLATIONS__LOAD_WARN",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Translations > Sagas matchLocaleSaga should match snapshot when a matching locale is found 1`] = `
Object {
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`Translations > Sagas useTranslationSaga should match snapshot 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "correlationId": "FAKE_CORRELATION_ID",
          "payload": Object {
            "formats": undefined,
            "locale": "en",
            "messages": Object {},
          },
          "type": "@@intl/UPDATE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;
